File: pyqt/PyRibbon/QxRibbonButtonGroup.py
#!/usr/bin/env python3
# Copyright (c) 2023 maminjie <canpool@163.com>
# SPDX-License-Identifier: GPL-3.0

from typing import List

from PyQt5.QtCore import Qt, QSize, pyqtSignal, QEvent
from PyQt5.QtGui import QActionEvent
from PyQt5.QtWidgets import QFrame, QAction, QMenu, QToolButton, QWidget, \
    QHBoxLayout, QSizePolicy, QWidgetAction

from .QxRibbonButton import RibbonButton
from .QxRibbonControls import RibbonSeparator


class RibbonButtonGroupItem:

    def __init__(self, *args):
        """
        RibbonButtonGroupItem()
        RibbonButtonGroupItem(a: QAction, w: QWidget, cw: bool)
        """
        if len(args) < 3:
            self.action: QAction = None
            self.widget: QWidget = None
            self.customWidget: bool = False
        else:
            self.action: QAction = args[0]
            self.widget: QWidget = args[1]
            self.customWidget: bool = args[2]

    def compare(self, *args) -> bool:
        """
        compare(action: QAction) -> bool
        compare(w: RibbonButtonGroupItem) -> bool
        """
        if len(args) < 1:
            return False
        if isinstance(args[0], QAction):
            return self.action == args[0]
        elif isinstance(args[0], RibbonButtonGroupItem):
            return self.action == args[0].action
        else:
            return False


class RibbonButtonGroup(QFrame):

    def __init__(self, parent=None):
        super().__init__(parent)
        self.m_items: List[RibbonButtonGroupItem] = list()

        layout = QHBoxLayout()
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(0)
        self.setLayout(layout)
        self.setSizePolicy(QSizePolicy.Maximum, QSizePolicy.Maximum)

    def addAction(self, *args) -> QAction:
        """
        addAction(a: QAction) -> QAction
        addAction(text: str, icon: QIcon, pop_mode=QToolButton.InstantPopup) -> QAction
        """
        if len(args) == 1:
            super().addAction(args[0])
            return args[0]
        else:
            a = QAction(args[1], args[0], self)
            super().addAction(a)
            pop_mode = QToolButton.InstantPopup if len(args) < 3 else args[2]
            if self.m_items:
                button: RibbonButton = self.m_items[-1].widget
                button.setPopupMode(pop_mode)
            return a

    def addMenu(self, menu: QMenu, pop_mode=QToolButton.InstantPopup) -> QAction:
        a = menu.menuAction()
        self.addAction(a)
        btn = self.m_items[-1].widget
        btn.setPopupMode(pop_mode)
        return a

    def addSeparator(self) -> QAction:
        a = QAction(self)
        a.setSeparator(True)
        self.addAction(a)
        return a

    def addWidget(self, w: QWidget):
        a = QWidgetAction(self)
        a.setDefaultWidget(w)
        w.setAttribute(Qt.WA_Hover)
        self.addAction(a)
        return a

    def hideWidget(self, action: QAction):
        i = len(self.m_items)
        for i, it in enumerate(self.m_items):
            if isinstance(it.action, QWidgetAction) and it.action == action:
                it.widget.hide()
                widgetAction: QWidgetAction = it.action
                widgetAction.releaseWidget(it.widget)
                break
        if i < len(self.m_items):
            self.m_items.pop(i)

    def sizeHint(self) -> QSize:
        return self.layout().sizeHint()

    def minimumSizeHint(self) -> QSize:
        return self.layout().minimumSize()

    def actionEvent(self, e: QActionEvent):
        item = RibbonButtonGroupItem()
        item.action = e.action()

        if e.type() == QEvent.ActionAdded:
            if isinstance(item.action, QWidgetAction):
                item.action.setParent(self)
                widgetAction: QWidgetAction = item.action
                """here widgetAction.requestWidget(self) is not widgetAction.defaultWidget()
                if using requestWidget will cause defaultWidget was deleted by python garbage collect
                see QWidgetAction source code, using defaultWidget() and set widget parent as self.
                """
                # item.widget = widgetAction.requestWidget(self)
                item.widget = widgetAction.defaultWidget()
                if item.widget:
                    item.widget.setParent(self)
                    item.widget.setAttribute(Qt.WA_LayoutUsesWidgetRect)
                    item.widget.show()
                    item.customWidget = True
            elif item.action.isSeparator():
                sp = RibbonSeparator(self)
                sp.setTopBottomMargins(3, 3)
                item.widget = sp
            if not item.widget:
                btn = RibbonButton(self)
                btn.setAutoRaise(True)
                btn.setFocusPolicy(Qt.NoFocus)
                btn.setButtonType(RibbonButton.SmallButton)
                btn.setToolButtonStyle(Qt.ToolButtonIconOnly)
                btn.setDefaultAction(item.action)
                btn.triggered.connect(self.actionTriggered)
                item.widget = btn
            self.layout().addWidget(item.widget)
            self.m_items.append(item)
        elif e.type() == QEvent.ActionChanged:
            self.layout().invalidate()
        elif e.type() == QEvent.ActionRemoved:
            # FIXME: whay clear all items?
            item.action.disconnect()
            for it in self.m_items:
                if isinstance(it.action, QWidgetAction) and it.customWidget:
                    widgetAction: QWidgetAction = it.action
                    widgetAction.releaseWidget(it.widget)
                else:
                    it.widget.hide()
                    it.widget.deleteLater()
            self.m_items.clear()
            self.layout().invalidate()

    # signals
    actionTriggered = pyqtSignal(QAction)

File: src/QxRibbonButtonGroup.cpp
/**
 * Copyleft (C) 2023 maminjie <canpool@163.com>
 * SPDX-License-Identifier: MIT
**/
#include "QxRibbonButtonGroup.h"
#include "QxRibbonControls.h"
#include "QxRibbonButton.h"

#include <QActionEvent>
#include <QHBoxLayout>
#include <QWidgetAction>

/* RibbonButtonGroupItem */
class RibbonButtonGroupItem
{
public:
    RibbonButtonGroupItem();
    RibbonButtonGroupItem(QAction *a, QWidget *w, bool cw);

    bool operator==(QAction *action);
    bool operator==(const RibbonButtonGroupItem &w);
public:
    QAction *action;
    QWidget *widget;
    bool customWidget;
};

RibbonButtonGroupItem::RibbonButtonGroupItem()
    : action(Q_NULLPTR)
    , widget(Q_NULLPTR)
    , customWidget(false)
{
}

RibbonButtonGroupItem::RibbonButtonGroupItem(QAction *a, QWidget *w, bool cw)
    : action(a)
    , widget(w)
    , customWidget(cw)
{
}

bool RibbonButtonGroupItem::operator==(QAction *action)
{
    return (this->action == action);
}

bool RibbonButtonGroupItem::operator==(const RibbonButtonGroupItem &w)
{
    return (this->action == w.action);
}

/* RibbonButtonGroupPrivate */
class RibbonButtonGroupPrivate
{
public:
    RibbonButtonGroupPrivate(RibbonButtonGroup *p);

    void init();
public:
    RibbonButtonGroup *q;
    QList<RibbonButtonGroupItem> m_items;
};

RibbonButtonGroupPrivate::RibbonButtonGroupPrivate(RibbonButtonGroup *p)
    : q(p)
{
}

void RibbonButtonGroupPrivate::init()
{
    QHBoxLayout *layout = new QHBoxLayout(q);
    layout->setContentsMargins(0, 0, 0, 0);
    layout->setSpacing(0);
    q->setSizePolicy(QSizePolicy::Maximum, QSizePolicy::Maximum);
}

/* RibbonButtonGroup */
RibbonButtonGroup::RibbonButtonGroup(QWidget *parent)
    : QFrame(parent)
    , d(new RibbonButtonGroupPrivate(this))
{
    d->init();
}

RibbonButtonGroup::~RibbonButtonGroup()
{
    for (RibbonButtonGroupItem &item : d->m_items) {
        if (QWidgetAction *widgetAction = qobject_cast<QWidgetAction *>(item.action)) {
            if (item.customWidget) {
                widgetAction->releaseWidget(item.widget);
            }
        }
    }
    delete d;
}

QAction *RibbonButtonGroup::addAction(QAction *a)
{
    QWidget::addAction(a);
    return a;
}

QAction *RibbonButtonGroup::addAction(const QString &text, const QIcon &icon,
                                      QToolButton::ToolButtonPopupMode popMode)
{
    QAction *a = new QAction(icon, text, this);

    addAction(a);
    RibbonButton *btn = qobject_cast<RibbonButton *>(d->m_items.last().widget);
    if (btn) {
        btn->setPopupMode(popMode);
    }
    return a;
}

QAction *RibbonButtonGroup::addMenu(QMenu *menu, QToolButton::ToolButtonPopupMode popMode)
{
    QAction *a = menu->menuAction();

    addAction(a);
    RibbonButton *btn = qobject_cast<RibbonButton *>(d->m_items.last().widget);
    if (btn) {
        btn->setPopupMode(popMode);
    }
    return a;
}

QAction *RibbonButtonGroup::addSeparator()
{
    QAction *a = new QAction(this);

    a->setSeparator(true);
    addAction(a);
    return (a);
}

QAction *RibbonButtonGroup::addWidget(QWidget *w)
{
    QWidgetAction *a = new QWidgetAction(this);

    a->setDefaultWidget(w);
    w->setAttribute(Qt::WA_Hover);
    addAction(a);
    return a;
}

QSize RibbonButtonGroup::sizeHint() const
{
    return layout()->sizeHint();
}

QSize RibbonButtonGroup::minimumSizeHint() const
{
    return layout()->minimumSize();
}

void RibbonButtonGroup::actionEvent(QActionEvent *e)
{
    RibbonButtonGroupItem item;
    item.action = e->action();

    switch (e->type()) {
    case QEvent::ActionAdded: {
        if (QWidgetAction *widgetAction = qobject_cast<QWidgetAction *>(item.action)) {
            widgetAction->setParent(this);
            item.widget = widgetAction->requestWidget(this);
            if (item.widget != Q_NULLPTR) {
                item.widget->setAttribute(Qt::WA_LayoutUsesWidgetRect);
                item.widget->show();
                item.customWidget = true;
            }
        } else if (item.action->isSeparator()) {
            RibbonSeparator *sp = new RibbonSeparator(this);
            sp->setTopBottomMargins(3, 3);
            item.widget = sp;
        }
        if (!item.widget) {
            RibbonButton *button = new RibbonButton(this);
            button->setAutoRaise(true);
            button->setFocusPolicy(Qt::NoFocus);
            button->setButtonType(RibbonButton::SmallButton);
            button->setToolButtonStyle(Qt::ToolButtonIconOnly);
            button->setDefaultAction(item.action);
            QObject::connect(button, &RibbonButton::triggered, this, &RibbonButtonGroup::actionTriggered);
            item.widget = button;
        }
        layout()->addWidget(item.widget);
        d->m_items.append(item);
    } break;
    case QEvent::ActionChanged: {
        layout()->invalidate();
    } break;
    case QEvent::ActionRemoved: {
        item.action->disconnect(this);
        auto i = d->m_items.begin();
        for (; i != d->m_items.end();) {
            QWidgetAction *widgetAction = qobject_cast<QWidgetAction *>(i->action);
            if ((widgetAction != 0) && i->customWidget) {
                widgetAction->releaseWidget(i->widget);
            } else {
                // destroy the QToolButton/QToolBarSeparator
                i->widget->hide();
                i->widget->deleteLater();
            }
            i = d->m_items.erase(i);
        }
        layout()->invalidate();
    } break;
    default:
        break;
    }
}
